src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs

```diff
@@ -103,15 +155,25 @@ internal static T[] ToArray<T>(this IEnumerable<T> sequence, int count)
             Requires.NotNull(sequence, nameof(sequence));
             Requires.Range(count >= 0, nameof(count));
 
+            if (count == 0)
+            {
+                return ImmutableArray<T>.Empty.array;
+            }
+
             T[] array = new T[count];
-            int i = 0;
-            foreach (var item in sequence)
+
+            if (!sequence.TryCopyTo(array, 0))
             {
-                Requires.Argument(i < count);
-                array[i++] = item;
+                int i = 0;
+                foreach (var item in sequence)
+                {
+                    Requires.Argument(i < count);
+                    array[i++] = item;
+                }
+
+                Requires.Argument(i == count);
             }
 
-            Requires.Argument(i == count);
```

Comment:
@AArnott, `i` is now only used when we're copying items one-by-one from an unknown enumerable. It doesn't make sense to do validation if `TryCopyTo` handles things. It's like trying to validate

```cs
var collection = source as ICollection<TSource>;
collection.CopyTo(array, 0);
// ??? How do we verify that the whole collection was copied over?
return array;
```